### PR TITLE
#14 Implement notification cancellation

### LINKED ISSUE #14: There is no API to delete notifications
There is no way to close a notification.

For example, there is no need to keep showing "unread messages" notification after messages are read, but the library can't stop showing it.

### PR DESCRIPTION
**Pull request by basilevs**
Fixes #14

This introduces a way for user to cancel previously sent notification.
The implementation is naive and dirty - no threading or processes involved, notification is cancelled by its ID.

[I've tried](https://github.com/basilevs/macos-notifications/tree/cancel) a "proper" solution with remote controlled singular side-car process, but found, that this is much simpler, as actual notification object is not needed to cancel it, and cancellation is a non-blocking operation.

Usage example:

```
from mac_notifications.client import create_notification, Notification 
mac_notification: Notification = None
def set_network_status(message):
    global mac_notification
    if mac_notification:
        mac_notification.cancel()
        mac_notification = None
    if message:
        mac_notification = create_notification(title="Network status", text=message)

set_network_status("Router is unavailable")
set_network_status(None) #Nominal
set_network_status("Domain name server is unavailable")

```

### CODE REVIEW

File: examples/cancel_multiple_notifications.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+from pathlib import Path
+
+from mac_notifications import client
+from mac_notifications.client import Notification
+
+
+if __name__ == "__main__":
+    print("Sending meeting notification.")
+
+    sent: [Notification] = []
```
Comment (Jorricks) [inline examples/cancel_multiple_notifications.py]: I don't think this is correct Python3.8 syntax.

```suggestion
    sent: List[Notification] = [] 
```

File: src/mac_notifications/notification_sender.py [hunk 2]
```diff
@@ -107,3 +106,9 @@ def userNotificationCenter_didActivateNotification_(
 
     # return notification
     return new_notif
+
+
+def cancel_notification(uid:str):
```
Comment (Jorricks) [inline src/mac_notifications/notification_sender.py]:
```suggestion
def cancel_notification(uid: str) -> None:
```

File: src/mac_notifications/manager.py [hunk 3]
```diff
@@ -30,6 +31,15 @@
 logger = logging.getLogger()
 
 
+class Notification(object):
+    def __init__(self, uid):
```
Comment (Jorricks) [inline src/mac_notifications/manager.py]:
```suggestion
    def __init__(self, uid: str) -> None:
```

File: src/mac_notifications/manager.py [hunk 4]
```diff
@@ -30,6 +31,15 @@
...
+        self.uid = uid
+
+    def cancel(self):
```
Comment (Jorricks) [inline src/mac_notifications/manager.py]:
```suggestion
    def cancel(self) -> None:
```

File: src/mac_notifications/manager.py [hunk 5]
```diff
@@ -12,6 +12,7 @@
 from mac_notifications.listener_process import NotificationProcess
 from mac_notifications.notification_config import NotificationConfig
 from mac_notifications.singleton import Singleton
+from .notification_sender import cancel_notification
```
Comment (Jorricks) [inline src/mac_notifications/manager.py]:
```suggestion
from mac_notifications import notification_sender
```

File: src/mac_notifications/manager.py [hunk 6]
```diff
@@ -30,6 +31,15 @@
 logger = logging.getLogger()
 
 
+class Notification(object):
+    def __init__(self, uid):
+        self.uid = uid
+
+    def cancel(self):
+        cancel_notification(self.uid)
```
Comment (Jorricks) [inline src/mac_notifications/manager.py:39]:
```suggestion
        notification_sender.cancel_notification(self.uid)
```

File: examples/cancel_multiple_notifications.py (hunk 1 above)
Comment (Jorricks) [reply examples/cancel_multiple_notifications.py]: We also need to add the `from typing import List`

**Comment by Jorricks** [comment]
Hello @basilevs,
Thanks a lot that you took the time to implement this, that's amazing!
I like your implementation, it's clear and easy. I just have some typing 'complaints', which I added as suggestions to this MR.
Let's get this out in a new version soon :v:

**Comment by basilevs** [comment]
>  some typing 'complaints', which I added as suggestions to this MR.

I don't see the suggestion make sure you've posted them (they are organized in reviews and are only published when the review is published)

**Review by Jorricks — COMMENTED** [review]
Woops. I use Gitlab during the day so sometimes it's a bit searching on Github. Thanks for the pointer :v:

File: examples/cancel_multiple_notifications.py (hunk 1 above)
Comment (basilevs) [reply examples/cancel_multiple_notifications.py]: Fixed with `list` builtin

File: src/mac_notifications/notification_sender.py (hunk 2 above)
Comment (basilevs) [reply src/mac_notifications/notification_sender.py]: Done

File: src/mac_notifications/manager.py (hunk 3 above)
Comment (basilevs) [reply src/mac_notifications/manager.py]: Done

File: src/mac_notifications/manager.py (hunk 4 above)
Comment (basilevs) [reply src/mac_notifications/manager.py]: Done

File: src/mac_notifications/manager.py (hunk 5 above)
Comment (basilevs) [reply src/mac_notifications/manager.py]: Replaced with

    from mac_notifications.notification_sender import cancel_notification

File: src/mac_notifications/manager.py (hunk 6 above)
Comment (basilevs) [reply src/mac_notifications/manager.py:39]: See above

**Comment by basilevs** [comment]
All suggestions have been acted upon

**Review by Jorricks — APPROVED** [review]
Looks good. Let's merge it :)

**Pull request merged**